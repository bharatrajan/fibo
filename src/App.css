.app{
  padding: 10%;
}

.error-msg-box-wrapper{
  height: 20px;
  font-size: 12px;
  color:red;
}

.input-box-wrapper{
  padding: 10px 0px;
}

.input-box-wrapper input{
  height: 20px;
  padding: 4px 6px;
  font-size: 14px;
  line-height: 20px;
  color: #555;
  -webkit-border-radius: 4px;
  -moz-border-radius: 4px;
  border-radius: 4px;  
  margin-right: 10px;
}
.submit-button{
  padding: 8px 20px;
  -webkit-border-radius: 4px;
  -moz-border-radius: 4px;
  border-radius: 4px;  
  color: #FFF;
  background-color: #04C;
}

.result-box-wrapper{
  padding: 10px 0px;  
  word-wrap: break-word;
  text-align: left;
}