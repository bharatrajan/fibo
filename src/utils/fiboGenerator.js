import bignumber from 'bignumber.js';

export const getFibo = (n, fiboStorage) => {
    let startTime = new Date().getTime();

    if(typeof fiboStorage[n] !== 'undefined'){
      let endTime = startTime - new Date().getTime();
      console.log(`${n} : Time taken = ${endTime}`)  
      return fiboStorage[n]
    };

    for(let i = fiboStorage.length; i < n+1 ; i++){ 
        fiboStorage[i] = fiboStorage[i-1].plus(fiboStorage[i-2]);
    }

    let endTime = new Date().getTime() - startTime;
    console.log(`${n} : Time taken = ${endTime}`)

    return fiboStorage[n];
}  
