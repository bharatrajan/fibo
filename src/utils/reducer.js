const trailingZeroFactory = {
    0: "",
    1: "0",
    2: "00",
    3: "000",
    4: "0000",
    5: "00000",
    6: "000000",
    7: "0000000",
    8: "00000000",
    9: "000000000",
    10: "0000000000",
    11: "00000000000",
    12: "000000000000",
    13: "0000000000000",
  }

export const resultReducer = arr => {
    let i, result = arr[0].toString();
    for(i = 1; i < arr.length; i++){
      let numOfZerosNeeded = 14 - arr[i].toString().length;
      result += trailingZeroFactory[numOfZerosNeeded] + arr[i]; 
    }
    return result;
}