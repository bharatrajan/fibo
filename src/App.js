import React, { Component } from 'react';
import logo from './logo.svg';
import './App.css';
import bigInt from 'big-integer';

class App extends Component {
  constructor(props) {
    super(props);
    this.onSubmit = this.onSubmit.bind(this);    
  }
  
  fiboArray = [bigInt(0), bigInt(1), bigInt(1)];

  state = {
    result : bigInt(0),
    isNValid : true
  }

  getFibo = n => {
    if(typeof this.fiboArray[n] !== 'undefined') return this.fiboArray[n];

    for(let i = this.fiboArray.length; i<n+1; i++){ 
      this.fiboArray[i] = this.fiboArray[i-1].add(this.fiboArray[i-2]);
      let str = `${this.fiboArray[i]} of ${i} is safe : ${Number.isSafeInteger(this.fiboArray[i])}`;
      console.log(str)
    }
    console.log("this.fiboArray : ", this.fiboArray);
    return this.fiboArray[n];
  }

  onSubmit = event => {
    const inputBoxVal = this.refs.inputBox.value;
    const n = parseInt(inputBoxVal);

    if(typeof n == 'number' && n > 0){
      let nthFibo = this.getFibo(n);
      let result = nthFibo.isSmall ? nthFibo.value : nthFibo.value.reverse().join(',')
      this.setState({
        result,
        isNValid : true
      });
    }else{
      this.setState({
        result: 0,
        isNValid : false    
      })
    }



  };  

  render() {
    const {result, isNValid} = this.state;
    return (
      <div>
        <div className="inputBoxWrapper">
          n = <input
                type="text"
                ref="inputBox"
                placeholder="Positive integer"/>
          {(!isNValid && 
            <span className="validityText"> Please give a positive integer </span> 
          )}      
        </div>
        
        <div className="buttonBoxWrapper">
          <button onClick={this.onSubmit}> GO </button>
        </div>  

        <div className="resultWrapper">     
          {(result == 0 ? "" : result)}
        </div>
     </div>
    );
  }
}

export default App;
