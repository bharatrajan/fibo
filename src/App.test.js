import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';

import App from './App';

import bignumber from 'bignumber.js';
import { resultReducer } from './utils/reducer.js';
import { getFibo, getFiboNCache } from './utils/fiboGenerator.js';


describe('set up', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
  });


  it('tests jest is there or not', ()=>{
    let cb = jest.fn();
        cb.mockReturnValueOnce(true)
    let ans = cb("My val")
    console.log("cb : ", cb.mock.calls)
    expect(ans).toBeTruthy();
  })  
})

describe('Select box', () => {
  it('tests select box default value', ()=>{
    const component = shallow(<App />);
    const selectBox = component.find('select');
    expect(selectBox.node.props.defaultValue).toEqual('audi');    
  });  

  it('tests select box onChange', ()=>{
    const component = shallow(<App />);
    const selectBox = component.find('select');
    let cb = jest.fn();
    selectBox.onChange = cb;

    selectBox.simulate('change', {target : { value : 'saab'}})
    console.log("cb.mock.calls[0][0] :", cb.mock.calls)
    //expect()
  })  
})

describe('comparison', () => {
  it('should match the snapshot data 003', () => {
    const component = renderer.create(<App maami="003"></App>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should match the snapshot data foo', () => {
    const component = renderer.create(<App maami="foo"></App>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });  

  it("input box placeholder='Positive integer'", () => {
    const component = shallow(<App />);
    const textBox = component.find('input');
    expect(textBox.node.props.placeholder).toEqual('Positive integer');
  });

  it('find timings with cache', () => {
    let fiboArray = [bignumber(0), bignumber(1), bignumber(1)];

    let result1 = getFiboNCache(1, fiboArray);
    let result2 = getFiboNCache(20, fiboArray);
    let result4 = getFiboNCache(50, fiboArray);
    let result3 = getFiboNCache(100, fiboArray);
    let result7 = getFiboNCache(200, fiboArray);
    let result8 = getFiboNCache(2004, fiboArray);
    let result9 = getFiboNCache(5050, fiboArray);
    let result10 = getFiboNCache(3055, fiboArray);
    let result11 = getFiboNCache(1035, fiboArray);
    let result12 = getFiboNCache(2003, fiboArray);
    let result5 = getFiboNCache(100000, fiboArray);

    expect(0).toEqual(0);
  });

  it('find timings without cache', () => {
    let fiboObj = {
      0: bignumber(0),
      1: bignumber(1),
      2: bignumber(1)
    };

    let result1 = getFibo(1, fiboObj);
    let result2 = getFibo(20, fiboObj);
    let result4 = getFibo(50, fiboObj);
    let result3 = getFibo(100, fiboObj);
    let result7 = getFibo(200, fiboObj);
    let result8 = getFibo(2004, fiboObj);
    let result9 = getFibo(5050, fiboObj);
    let result10 = getFibo(3055, fiboObj);
    let result11 = getFibo(1035, fiboObj);
    let result12 = getFibo(2003, fiboObj);
    let result5 = getFibo(100000, fiboObj);

    expect(0).toEqual(0);
  });
});

