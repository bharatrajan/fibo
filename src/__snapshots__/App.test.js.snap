// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`comparison should match the snapshot data 1`] = `
<div
  className="app"
>
  1
  <div
    className="input-box-wrapper"
  >
    n =
     
    <input
      placeholder="Positive integer"
      type="text"
    />
    <select
      defaultValue="saab"
      onChange={[Function]}
    >
      <option
        value="volvo"
      >
        Volvo
      </option>
      <option
        value="saab"
      >
        Saab
      </option>
      <option
        value="mercedes"
      >
        Mercedes
      </option>
      <option
        value="audi"
      >
        Audi
      </option>
    </select>
    <button
      className="submit-button"
      onClick={[Function]}
    >
       
      Submit
       
    </button>
  </div>
  <div
    className="error-msg-box-wrapper"
  />
  <div
    className="result-box-wrapper"
  >
    
  </div>
</div>
`;
